File: src/org/mapsforge/map/layer/renderer/CanvasRastererEx.java
package org.mapsforge.map.layer.renderer;

import org.jetbrains.annotations.NotNull;
import org.mapsforge.core.graphics.*;
import org.mapsforge.core.mapelements.MapElementContainer;
import org.mapsforge.core.model.Point;
import org.mapsforge.core.model.Tile;

import java.util.Collection;
import java.util.List;

public class CanvasRastererEx {
  private final Canvas canvas;
  private final Matrix symbolMatrix;

  CanvasRastererEx(@NotNull GraphicFactory graphicFactory) {
    canvas = graphicFactory.createCanvas();
    symbolMatrix = graphicFactory.createMatrix();
  }

  static void drawWays(List<List<ShapePaintContainer>>[] drawWays, Shape shape) {
    int levelsPerLayer = drawWays[0].size();
    Paint currentStroke = null;
    for (List<List<ShapePaintContainer>> shapePaintContainers : drawWays) {
      for (int level = 0; level < levelsPerLayer; level++) {
        List<ShapePaintContainer> wayList = shapePaintContainers.get(level);
        for (int i = wayList.size() - 1; i >= 0; i--) {
          ShapePaintContainer container = wayList.get(i);

          assert container.paint == null;
          Paint fill = container.fill;
          boolean addEndFill = false;
          if (fill != null && !fill.isTransparent()) {
            addEndFill = shape.beginFillOrSetLineStyle(fill);
            assert addEndFill;
          }

          if (currentStroke != container.stroke) {
            currentStroke = container.stroke;
            if (currentStroke != null && !currentStroke.isTransparent()) {
              boolean r = shape.beginFillOrSetLineStyle(currentStroke);
              assert !r;
            }
          }

          drawShapePaintContainer(container, shape);

          if (addEndFill) {
            shape.endFill();
          }
        }
      }
    }
  }

  void drawMapElements(Collection<MapElementContainer> elements, Tile tile) {
    Point origin = tile.getOrigin();
    for (MapElementContainer element : elements) {
      element.draw(canvas, origin, symbolMatrix);
    }
  }

  void fill(int color) {
    if (GraphicUtils.getAlpha(color) > 0) {
      canvas.fillColor(color);
    }
  }

  void setCanvasBitmap(Bitmap bitmap) {
    canvas.setBitmap(bitmap);
  }

  private static void drawCircleContainer(@NotNull ShapePaintContainer shapePaintContainer, @NotNull Shape shape) {
    CircleContainer circleContainer = (CircleContainer)shapePaintContainer.shapeContainer;
    Point point = circleContainer.point;
    shape.drawCircle(point.x, point.y, circleContainer.radius);
  }

  private static void drawPath(@NotNull ShapePaintContainer shapePaintContainer, float dy, @NotNull Shape shape) {
    PolylineContainer shapeContainer = (PolylineContainer)shapePaintContainer.shapeContainer;
    Point[][] coordinates = dy == 0 ? shapeContainer.getCoordinatesAbsolute() : shapeContainer.getCoordinatesRelativeToTile();
    Point tileOrigin = dy == 0 ? shapeContainer.getTile().getOrigin() : new Point(0, 0);
    for (Point[] innerList : coordinates) {
      Point[] points = dy == 0 ? innerList : RendererUtils.parallelPath(innerList, dy);
      if (points.length >= 2) {
        shape.drawPolyLine(points, tileOrigin);
      }
    }
  }

  private static void drawShapePaintContainer(@NotNull ShapePaintContainer shapePaintContainer, @NotNull Shape shape) {
    switch (shapePaintContainer.shapeContainer.getShapeType()) {
      case CIRCLE:
        drawCircleContainer(shapePaintContainer, shape);
        break;

      case POLYLINE:
        drawPath(shapePaintContainer, shapePaintContainer.dy, shape);
        break;
    }
  }
}

File: mapsforge-ex/src/CanvasRastererEx.kt
package org.mapsforge.map.layer.renderer

import org.mapsforge.core.graphics.Paint
import org.mapsforge.core.mapelements.MapElementContainer
import org.mapsforge.core.mapelements.SymbolContainer
import org.mapsforge.core.mapelements.WayTextContainer
import org.mapsforge.core.model.Point
import org.mapsforge.core.model.Tile

val EMPTY_POINT:Point = Point(0.0, 0.0)

public trait MapElementContainerEx {
  fun draw(canvas:CanvasEx, origin:Point):Unit
}

fun drawWays(drawWays:Array<List<MutableList<ShapePaintContainer>>>, canvas:CanvasEx) {
  val levelsPerLayer = drawWays[0].size()
  var currentStroke:Paint? = null
  for (shapePaintContainers in drawWays) {
    for (level in 0..levelsPerLayer - 1) {
      val wayList = shapePaintContainers.get(level)
      for (i in wayList.size() - 1 downTo 0) {
        val container = wayList.get(i)

        assert(container.paint == null)
        val fill = container.fill
        var addEndFill = false
        if (fill != null && !fill.isTransparent()) {
          addEndFill = canvas.beginFillOrSetLineStyle(fill)
          assert(addEndFill)
        }

        if (currentStroke != container.stroke) {
          currentStroke = container.stroke
          if (currentStroke != null && !currentStroke!!.isTransparent()) {
            val r = canvas.beginFillOrSetLineStyle(currentStroke!!)
            assert(!r)
          }
        }

        drawShapePaintContainer(container, canvas)

        if (addEndFill) {
          canvas.endFill()
        }
      }
    }
  }
}

private fun drawShapePaintContainer(shapePaintContainer:ShapePaintContainer, canvas:CanvasEx) {
  when (shapePaintContainer.shapeContainer.getShapeType()) {
    ShapeType.CIRCLE -> {
      val circleContainer = shapePaintContainer.shapeContainer as CircleContainer
      val point = circleContainer.point
      canvas.drawCircle(point.x, point.y, circleContainer.radius)
    }

    ShapeType.POLYLINE -> {
      val shapeContainer = shapePaintContainer.shapeContainer as PolylineContainer
      val coordinates = if (shapePaintContainer.dy == 0f) shapeContainer.getCoordinatesAbsolute() else shapeContainer.getCoordinatesRelativeToTile()
      [suppress("CAST_NEVER_SUCCEEDS")]
      canvas.drawPolyLine(coordinates, if (shapePaintContainer.dy == 0f) shapeContainer.getTile().getOrigin() else EMPTY_POINT, shapePaintContainer.dy)
    }
    else -> {
    }
  }
}

fun drawMapElements(elements:Collection<MapElementContainer>, tile:Tile, canvas:CanvasEx) {
  val origin = tile.getOrigin()
  for (element in elements) {
    if (element is WayTextContainer) {
      canvas.drawTextRotated(element.text, element.getPoint(), element.end, origin, element.paintFront)
    }
    else if (element is SymbolContainer) {
      val boundary = element.getBoundary()
      val point = element.getPoint()
      if (element.theta != 0f) {
        throw UnsupportedOperationException("rotated symbol not supported")
      }
      canvas.drawSymbol(element.symbol, (point.x - origin.x) + boundary.left, (point.y - origin.y) + boundary.top, element.theta)
    }
    else if (element is MapElementContainerEx) {
      element.draw(canvas, origin);
    }
    else {
      throw UnsupportedOperationException("unsupported element")
    }
  }
}